std/container/array.d

```diff
@@ -59,87 +48,255 @@ if (!is(Unqual!T == bool))
-                 * than realloc.
-                 */
-                immutable oldLength = length;
-                auto newPayload =
-                    enforce(cast(T*) malloc(sz))[0 .. oldLength];
-                // copy old data over to new array
-                memcpy(newPayload.ptr, _payload.ptr, T.sizeof * oldLength);
-                // Zero out unused capacity to prevent gc from seeing
-                // false pointers
-                memset(newPayload.ptr + oldLength,
-                        0,
-                        (elements - oldLength) * T.sizeof);
-                GC.addRange(newPayload.ptr, sz);
-                GC.removeRange(_payload.ptr);
-                free(_payload.ptr);
-                _payload = newPayload;
+                // Move elements over by one slot
+                memmove(_payload.ptr + index,
+                        _payload.ptr + index + 1,
+                        T.sizeof * (length - index));
+                
+                // Zero out unused capacity to prevent gc from seeing false pointers
+                _payload[length] = T.init;
+            }
+        }
+        
+        // Remove a range of entries starting at the specified index.
+        // An entry will be destroyed appropriately.
+        //
+        // Complexity: $(BIGOH n)
+        void removeAt(size_t index, size_t howMany)
+        {
+            enforce(index >= 0 && howMany > 0);
+            enforce(index + howMany >= 0);
+            enforce(index + howMany <= length);
+            
+            static if (is(T == struct) && hasElaborateDestructor!T)
+            {
+                foreach (immutable n; index .. index + howMany)
+                    .destroy(_payload[n]); // Destroy this guy
+            }
+            
+            immutable tailCount = length - (index + howMany);
+            if (tailCount > 0)
+            {
+                memmove(_payload.ptr + index,
+                        _payload.ptr + (index + howMany),
+                        T.sizeof * tailCount);
+                initializeAll(_payload.ptr[(length - howMany) .. (length - howMany) + howMany]);
             }
             else
             {
-                /* These can't have pointers, so no need to zero
-                 * unused region
-                 */
-                auto newPayload =
-                    enforce(cast(T*) realloc(_payload.ptr, sz))[0 .. length];
-                _payload = newPayload;
+                initializeAll(_payload.ptr[index .. index + howMany]);
+            }
+            
+            _payload = _payload.ptr[0 .. _payload.length - howMany];
+        }
+
+        // Insert an entry at the specified index.
+        //
+        // Returns the number of entries inserted into the Array.
+        //
+        // Complexity: $(BIGOH n)
+        size_t insertAt(Stuff)(size_t index, Stuff stuff)
+            if (isImplicitlyConvertible!(Stuff, T))
+        {
+            enforce(index <= length);
```

Comment:
Please don't use enforce in library code. enforce throws `exception` and should only be used for things that can "legitimately fail". If a request for an insertion happens at an index bigger than length, then it's a programming error. `assert` is called for in these cases.